probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -0,0 +1,815 @@
+                csw.AddrInc = AddressIncrement::Off;
+                csw.Size = DataSize::U8;
+
+                csw.Instruction = false;
+                // Enable secure access if it's allowed
+                csw.NonSecure = !csw.SPIDEN;
+                csw.Privileged = true;
+                csw.CACHE = 0;
+                csw.MTE = false;
+
+                CSW::try_from(Into::<u32>::into(csw))?
+            }
+        };
+
+        Ok(Box::new(BlackMagicProbeMemoryInterface {
+            probe: self,
+            current_ap,
+            index: 0,
+            apsel: 0,
+            csw: csw.into(),
+        }) as _)
+    }
+
+    fn read_chip_info_from_rom_table(
+        &mut self,
+        dp: DpAddress,
+    ) -> Result<Option<crate::architecture::arm::ArmChipInfo>, ArmError> {
+        if dp != DpAddress::Default {
+            return Err(ArmError::NotImplemented("multidrop not yet implemented"));
+        }
+
+        for ap in self.access_ports.clone() {
+            if let Ok(mut memory) = self.memory_interface(&ap) {
+                let base_address = memory.base_address()?;
+                let component = Component::try_parse(&mut *memory, base_address)?;
+
+                if let Component::Class1RomTable(component_id, _) = component {
+                    if let Some(jep106) = component_id.peripheral_id().jep106() {
+                        return Ok(Some(ArmChipInfo {
+                            manufacturer: jep106,
+                            part: component_id.peripheral_id().part(),
+                        }));
+                    }
+                }
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+impl SwoAccess for BlackMagicProbeArmDebug {
+    fn enable_swo(&mut self, config: &crate::architecture::arm::SwoConfig) -> Result<(), ArmError> {
+        println!("Enabling SWO: {:?}", config);
+        unimplemented!()
+    }
+
+    fn disable_swo(&mut self) -> Result<(), ArmError> {
+        println!("Disabling SWO");
+        unimplemented!()
+    }
+
+    fn read_swo_timeout(&mut self, timeout: std::time::Duration) -> Result<Vec<u8>, ArmError> {
+        println!("Setting read swo timeout: {:?}", timeout);
+        unimplemented!()
+    }
+}
+
+impl SwdSequence for BlackMagicProbeArmDebug {
+    fn swj_sequence(
+        &mut self,
+        bit_len: u8,
+        bits: u64,
+    ) -> Result<(), crate::probe::DebugProbeError> {
+        self.probe.swj_sequence(bit_len, bits)
+    }
+
+    fn swj_pins(
+        &mut self,
+        pin_out: u32,
+        pin_select: u32,
+        pin_wait: u32,
+    ) -> Result<u32, crate::probe::DebugProbeError> {
+        self.probe.swj_pins(pin_out, pin_select, pin_wait)
+    }
+}
+
+fn dp_to_bmp(dp: DpAddress) -> Result<u8, ArmError> {
+    match dp {
+        DpAddress::Default => Ok(0),
+        DpAddress::Multidrop(val) => val
+            .try_into()
+            .map_err(|_| ArmError::NotImplemented("multidrop value is out of range")),
```

Comment:
NotImplemented is for functionality that's not implemented. If there is no fitting variant we can introduce one :)

Reply:
What should I return if it's supported but I haven't implemented it yet?

Reply:
Oh, it can support larger than u8 values? I guess I misunderstood then :)

Small nit: Can we make the error message tell the user what the allowed range of values is and what value was received? :)

Reply:
Sorry, I'd forgotten what this code was doing. It can only support `u8` values. If you had an extremely long JTAG chain with more than 256 TAPs, then this value would be exceeded. Would I use `OutOfBounds`?

Reply:
That sounds correct, let's go with it! Truth be told, I am not the most familiar with JTAG :)